File: contents/cooley_tukey/code/nim/fft.nim
from math import PI
from random import rand, randomize
from sequtils import map, toSeq, zip
from sugar import `=>`
import complex
# $ nimble install fftw3
import fftw3

# For some reason this isn't in the Nim fftw3 bindings.
const FFTW_FORWARD = -1

proc allClose[T: SomeNumber | Complex[SomeFloat]](reference: openArray[T], calculated: openArray[T], thresh = 1.0e-13): bool =
  if reference.len != calculated.len:
    return false
  result = true
  for (r, c) in zip(reference, calculated):
    if abs(r - c) > thresh:
      result = false
      break

proc skip[T](inp: openArray[T], stride: int): seq[T] =
  ## Like `inp[::stride]` in Python.
  if stride < 1:
    return toSeq(inp)
  result = @[]
  var i = 0
  while i < inp.len:
    result.add(inp[i])
    i += stride

proc nim_to_fftw(inp: Complex64): fftw_complex =
  [inp.re, inp.im]

proc fftw_to_nim(inp: fftw_complex): Complex64 =
  complex64(inp[0], inp[1])

proc nim_to_fftw(inp: openArray[Complex64]): seq[fftw_complex] =
  inp.map(v => nim_to_fftw(v))

proc fftw_to_nim(inp: openArray[fftw_complex]): seq[Complex64] =
  inp.map(v => fftw_to_nim(v))

proc dft_fftw3(inp: openArray[Complex64]): seq[Complex64] =
  var
    fftw_inp = nim_to_fftw(inp)
    fftw_out = newSeq[fftw_complex](inp.len)
  let p = fftw_plan_dft_1d(cint(inp.len),
                           addr(fftw_inp[low(fftw_inp)]),
                           addr(fftw_out[low(fftw_out)]),
                           FFTW_FORWARD,
                           FFTW_ESTIMATE)
  fftw_execute(p)
  fftw_destroy_plan(p)
  fftw_to_nim(fftw_out)

proc dft[T: SomeFloat](x: openArray[Complex[T]]): seq[Complex[T]] =
  let n = x.len
  result = newSeq[Complex[T]](n)
  for i in 0..n - 1:
    for k in 0..n - 1:
      result[i] += x[k] * exp(-complex(0.0, 2.0) * PI * float(i * k / n))

proc cooley_tukey[T: SomeFloat](x: openArray[Complex[T]]): seq[Complex[T]] =
  let n = x.len
  if n <= 1:
    return toSeq(x)
  result = newSeq[Complex[T]](n)
  let
    even = cooley_tukey(skip(x, 2))
    odd = cooley_tukey(skip(x[1..high(x)], 2))
    midpoint = n div 2
  for k in 0..midpoint - 1:
    let exp_term = exp(-complex(0.0, 2.0) * PI * float(k / n)) * odd[k]
    result[k] = even[k] + exp_term
    result[k + midpoint] = even[k] - exp_term

# proc bit_reverse[T: SomeNumber](x: openArray[T]): seq[T]

# proc iterative_cooley_tukey[T: SomeFloat](x: openArray[T]): seq[Complex[T]] =
#   let n = x.len
#   result = bit_reverse(x)

when isMainModule:
  randomize()
  let
    x = toSeq(1..64).map(i => complex64(rand(1.0), 0.0))
    y = cooley_tukey(x)
    # z = iterative_cooley_tukey(x)
    t = dft(x)
    reference = dft_fftw3(x)
  assert(allClose(reference, y))
  # assert(allClose(reference, z))
  assert(allClose(reference, t))

File: contents/contents/cooley_tukey/code/c/fft.c
#include <complex.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <fftw3.h>

void fft(double complex *x, int n) {
    double complex y[n];
    memset(y, 0, sizeof(y));
    fftw_plan p;

    p = fftw_plan_dft_1d(n, (fftw_complex*)x, (fftw_complex*)y,
                         FFTW_FORWARD, FFTW_ESTIMATE);

    fftw_execute(p);
    fftw_destroy_plan(p);

    for (size_t i = 0; i < n; ++i) {
        x[i] = y[i] / sqrt((double)n);
    }
}

void dft(double complex *X, const size_t N) {
    double complex tmp[N];
    for (size_t i = 0; i < N; ++i) {
        tmp[i] = 0;
        for (size_t j = 0; j < N; ++j) {
            tmp[i] += X[j] * cexp(-2.0 * M_PI * I * j * i / N);
        }
    }

    memcpy(X, tmp, N * sizeof(*X));
}

void cooley_tukey(double complex *X, const size_t N) {
    if (N >= 2) {
        double complex tmp [N / 2];
        for (size_t i = 0; i < N / 2; ++i) {
            tmp[i] = X[2*i + 1];
            X[i] = X[2*i];
        }
        for (size_t i = 0; i < N / 2; ++i) {
            X[i + N / 2] = tmp[i];
        }

        cooley_tukey(X, N / 2);
        cooley_tukey(X + N / 2, N / 2);

        for (size_t i = 0; i < N / 2; ++i) {
            X[i + N / 2] = X[i] - cexp(-2.0 * I * M_PI * i / N) * X[i + N / 2];
            X[i] -= (X[i + N / 2]-X[i]);
        }
    }
}

void bit_reverse(double complex *X, size_t N) {
    for (int i = 0; i < N; ++i) {
        int n = i;
        int a = i;
        int count = (int)log2((double)N) - 1;

        n >>= 1;
        while (n > 0) {
            a = (a << 1) | (n & 1);
            count--;
            n >>= 1;
        }
        n = (a << count) & ((1 << (int)log2((double)N)) - 1);

        if (n > i) {
            double complex tmp = X[i];
            X[i] = X[n];
            X[n] = tmp;
        }
    }
}

void iterative_cooley_tukey(double complex *X, size_t N) {
    bit_reverse(X, N);

    for (int i = 1; i <= log2((double)N); ++i) {
        int stride = pow(2, i);
        double complex w = cexp(-2.0 * I * M_PI / stride);
        for (size_t j = 0; j < N; j += stride) {
            double complex v = 1.0;
            for (size_t k = 0; k < stride / 2; ++k) {
                X[k + j + stride / 2] = X[k + j] - v * X[k + j + stride / 2];
                X[k + j] -= (X[k + j + stride / 2] - X[k + j]);
                v *= w;
            }
        }
    }
}

void approx(double complex *X, double complex *Y, size_t N) {
    for (size_t i = 0; i < N; ++i) {
        if (cabs(X[i]) - cabs(Y[i]) > 1E-5) {
            printf("This is not approximate.\n");
            return;
        }
    }
    printf("This is approximate.\n");
}

int main() {
    srand(time(NULL));
    double complex x[64], y[64], z[64];
    for (size_t i = 0; i < 64; ++i) {
        x[i] = rand() / (double) RAND_MAX;
        y[i] = x[i];
        z[i] = x[i];
    }

    fft(x, 64);
    cooley_tukey(y, 64);
    iterative_cooley_tukey(z, 64);

    approx(x, y, 64);
    approx(x, z, 64);

    return 0;
}
